### PR TITLE
Fix High Score Music after winning scenario

### PR DESCRIPTION
**Pull request by zenseii**
This plays the Victory theme when winning a scenario after the Dragon animation like OG, instead of the Main Menu theme like  it does currently, and it makes sure that the Main Menu theme continues to play if opening the High Score from the Main Menu, like in the original game.

This also future-proofs for when the Victory theme will be played while showing the Congratulations text behind the painting after the Dragon animation, but before the High Score menu is displayed.

### CODE REVIEW

File: src/fheroes2/game/game_startgame.cpp [hunk 1]
```diff
@@ -646,7 +646,12 @@ fheroes2::GameMode Interface::Basic::StartGame()
 
                     res = HumanTurn( loadedFromSave );
 
-                    // reset environment sounds and music theme at the end of the human turn
+                    // Audio has already been reset if scenario is won and WIN.SMK has played.
+                    if ( Game::CurrentMusic() == MUS::VICTORY ) {
```
Comment (oleg-derevenetz) [inline src/fheroes2/game/game_startgame.cpp:650]: Hi @zenseii there is a race condition. If `AGG::PlayMusic()` is called in async mode, then `Game::SetCurrentMusic()` will be called in the background thread, and there is no guarantee that it will be set to the `MUS::VICTORY` here.

Comment (zenseii) [reply src/fheroes2/game/game_startgame.cpp:650]: Thank you for your comment @oleg-derevenetz ! You're absolutely right. I initially thought of using the `fheroes2::GameMode` value of `res` as a condition, but it would become more complex. So the simplest approach would be to set run async to false in this case.

The best would maybe be to run `HighScores()` after `PlayMusic( MUS::VICTORY)` to avoid checking this condition I wrote before ending every single turn, but this presents it's own set of issues when leaving the high scores menu.

Comment (oleg-derevenetz) [reply src/fheroes2/game/game_startgame.cpp:650]: Why just not check that if `res` is `GameMode::HIGHSCORES`, then we do not reset sounds? Like this:

```cpp
// Audio has already been reset if scenario is won and WIN.SMK has played.
if ( res == fheroes2::GameMode::HIGHSCORES) {
  break;
}
```

Please also note that `WIN.SMK` could be played (and highscores screen is shown) when you won a campaign, this is handled in different place in the campaign-related code.

Comment (zenseii) [reply src/fheroes2/game/game_startgame.cpp:650]: Excuse me, I just saw your comment after commiting. Yes, the only issue here is that we could have `res = fheroes2::GameMode::Cancel` in the case of using the experimental option to continue playing the game after winning which is at the end of this if statement below.

```
                    AGG::PlayMusic( MUS::VICTORY, true, false );

                    res = fheroes2::GameMode::HIGHSCORES;

                    if ( conf.ExtGameContinueAfterVictory() && myKingdom.isPlay() ) {
                        if ( Dialog::YES == Dialog::Message( "", _( "Do you wish to continue the game?" ), Font::BIG, Dialog::YES | Dialog::NO ) ) {
                            continueAfterVictory = true;
                            // Game::HighScores() calls ResetResult()
                            Game::HighScores();
                            Interface::Basic & I = Interface::Basic::Get();
                            I.ResetFocus( GameFocus::HEROES );
                            I.SetRedraw( Interface::REDRAW_ALL );
                            res = fheroes2::GameMode::CANCEL;
                        }
                    }
 ```

In the case of `DIALOG::YES `we will still see the High Score before continuing to play the map, this would mean that the music would be reset since `res` would now be set to `fheroes2::GameMode::CANCEL`.

Comment (oleg-derevenetz) [reply src/fheroes2/game/game_startgame.cpp:650]: @zenseii If we continue the game after victory, this means that we do not exit to the main menu after showing the highscores (and we remain on the adventure map), so there is no reason to continue to play the `MUS::VICTORY`.

Comment (zenseii) [reply src/fheroes2/game/game_startgame.cpp:650]: @oleg-derevenetz Yes, so it would still be beneficial to run `ResetAudio()` after exiting the highscore if we continue to play, but the problem is that the music would be reset right before opening the High Scores and there would be no music if you choose to continue playing.

Comment (oleg-derevenetz) [reply src/fheroes2/game/game_startgame.cpp:650]: @zenseii

> but the problem is that the music would be reset right before opening the High Scores

Why? If we choose to continue to play, then highscores screen will be show immediately in the same piece of code you quoted above.

> and there would be no music if you choose to continue playing

There will be the terrain music on the adventure map as usual.

Comment (zenseii) [reply src/fheroes2/game/game_startgame.cpp:650]: > @zenseii
> 
> > but the problem is that the music would be reset right before opening the High Scores
> 
> Why? If we choose to continue to play, then highscores screen will be shown immediately in the same piece of code you quoted above.


There would be no music because at this point  `ResetAudio()` would be called immediately after when going back to `game_startgame.cpp` as seen below this comment.
> > and there would be no music if you choose to continue playing
> 
> There will be the terrain music on the adventure map as usual.

Yes, I realized my comment was ambiguous so I edited it. The adventure map music will indeed make sure that new audio is played, the problem is as stated above.

Comment (oleg-derevenetz) [reply src/fheroes2/game/game_startgame.cpp:650]: BTW if the problem is just that in the highscores screen should be played the `MUS::VICTORY` instead of the `MUS::MAINMENU`, then the change would be a one-liner in the `game_highscores.cpp`:

```cpp
AGG::PlayMusic( ( gameResult.GetResult() & GameOver::WINS ) ? MUS::VICTORY : MUS::MAINMENU, true, true );
```

Wouldn't this be enough?

Comment (oleg-derevenetz) [reply src/fheroes2/game/game_startgame.cpp:650]: @zenseii 

> There would be no music because at this point ResetAudio() would be called immediately after when going back to game_startgame.cpp as seen below this discussion.

Yes, but the `Game::HighScores()` will be called first.

Comment (zenseii) [reply src/fheroes2/game/game_startgame.cpp:650]: > BTW if the problem is just that in the highscores screen should be played the `MUS::VICTORY` instead of the `MUS::MAINMENU`, then the change would be a one-liner in the `game_highscores.cpp`:
> 
> ```c++
> AGG::PlayMusic( ( gameResult.GetResult() & GameOver::WINS ) ? MUS::VICTORY : MUS::MAINMENU, true, true );
> ```
> 
> Wouldn't this be enough?

@oleg-derevenetz Yes, this was my intial thought too, but this turned out to not be future-proof for when we will implement the correct behavior for the Dragon Animation, because `MUS::VICTORY` starts playing when you have the last frame of the Dragon animation still showing and the High Scores has not been shown yet and you are inputting name, as can be seen in the video in this issue:
https://github.com/ihhub/fheroes2/issues/2918

Basically ``MUS::VICTORY`` needs to be played before the High Scores menu.

Comment (zenseii) [reply src/fheroes2/game/game_startgame.cpp:650]: > @zenseii
> 
> > There would be no music because at this point ResetAudio() would be called immediately after when going back to game_startgame.cpp as seen below this discussion.
> 
> Yes, but the `Game::HighScores()` will be called first.

Yes, actually you are right. I don't know what I was thinking here but using `res == fheroes2::GameMode::HIGHSCORES` as the condition solves the problem. I just tested it. 
Thank you for you patience and explanation.

File: src/fheroes2/game/game_campaign.cpp [hunk 2]
```diff
@@ -816,6 +816,7 @@ fheroes2::GameMode Game::CompleteCampaignScenario( const bool isLoadingSaveFile
 
         AGG::ResetAudio();
         Video::ShowVideo( "WIN.SMK", Video::VideoAction::WAIT_FOR_USER_INPUT );
+        AGG::PlayMusic( MUS::VICTORY, true, true );
```
Comment (oleg-derevenetz) [inline src/fheroes2/game/game_campaign.cpp:819]: BTW we can add the TODO comment here with the brief explanation why the `AGG::PlayMusic()` is called here and that it should be replaced by a more complex approach in future.

File: src/fheroes2/game/game_over.cpp [hunk 3]
```diff
@@ -343,6 +343,7 @@ fheroes2::GameMode GameOver::Result::LocalCheckGameOver()
                 else {
                     AGG::ResetAudio();
                     Video::ShowVideo( "WIN.SMK", Video::VideoAction::WAIT_FOR_USER_INPUT );
+                    AGG::PlayMusic( MUS::VICTORY, true, true );
```
Comment (oleg-derevenetz) [inline src/fheroes2/game/game_over.cpp:346]: And here.

File: src/fheroes2/game/game_campaign.cpp (hunk 2 above)
Comment (zenseii) [reply src/fheroes2/game/game_campaign.cpp:819]: Yes, great idea 👍

**Review by ihhub — APPROVED** [review]

File: src/fheroes2/game/game_over.cpp (hunk 3 above)
Comment (zenseii) [reply src/fheroes2/game/game_over.cpp:349]: Done!

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by ihhub** [comment]
@zenseii , huge thanks for fixing this bug!
